ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/usermanagement/PersistentUserStoreTests.scala

```diff
@@ -0,0 +1,159 @@
+      val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(1))
+
+      val eventsQueue = new ConcurrentLinkedQueue[String]()
+      // Latches to ensure the initial state
+      val barStarted = new CountDownLatch(1)
+      val updateStatementInADone = new CountDownLatch(1)
+      val externalCheckDone = new CountDownLatch(1)
+      val fooIssuedUpdateQuery = new CountDownLatch(1)
+      val barIsAboutToIssueUpdateQuery = new CountDownLatch(1)
+
+      val userId = Ref.UserId.assertFromString("user1")
+      for {
+        // create a user with a known resource version
+        user <- inTransaction { connection =>
+          UserManagementStorageBackendImpl.createUser(
+            DbUserPayload(
+              id = userId,
+              primaryPartyO = None,
+              isDeactivated = false,
+              resourceVersion = 100,
+              createdAt = 0,
+            )
+          )(connection)
+          UserManagementStorageBackendImpl.getUser(userId)(connection).value
+        }
+        getUserFun = (connection: Connection) => {
+          UserManagementStorageBackendImpl
+            .getUser(user.payload.id)(connection)
+            .value
+            .payload
+            .resourceVersion
+        }
+        _ = user.payload.resourceVersion shouldBe 100
+        txA = Future {
+          barStarted.await()
+        }(ec).flatMap(_ =>
+          inTransaction { connection =>
+            eventsQueue.add("Foo0")
+            val updateSucceeded = updateRowFun(user.internalId, 100)(connection)
+            fooIssuedUpdateQuery.countDown()
+            val resourceVersion = getUserFun(connection)
+            barIsAboutToIssueUpdateQuery.await()
+            eventsQueue.add("Foo1")
+            updateStatementInADone.countDown()
+            externalCheckDone.await()
+            resourceVersion -> updateSucceeded
+          }
+        )(ec)
+        txB = inTransaction { connection =>
+          eventsQueue.add("Bar0")
+          barStarted.countDown()
+          // NOTE: This is the only countdown latch Bar is waiting on
+          fooIssuedUpdateQuery.await()
+          eventsQueue.add("Bar1")
+          barIsAboutToIssueUpdateQuery.countDown()
+          val updateSucceeded = updateRowFun(user.internalId, 100)(connection)
+          eventsQueue.add("Bar2")
+          val resourceVersion = getUserFun(connection)
+          resourceVersion -> updateSucceeded
+        }
+        _ = updateStatementInADone.await()
+        // NOTE: Sequence is of events {"Bar0", "Foo0", "Bar1"} represents an initial state guaranteed by the countdown latch setup.
+        //       It means both Foo and Bar transaction are in progress and Foo additionally completed an update query.
+        // NOTE: The thing we are testing for is that Bar cannot finish before Foo, because
+        //       both wrote to the same row and Foo did it faster.
+        _ = eventsQueue.iterator().asScala.toList shouldBe List("Bar0", "Foo0", "Bar1", "Foo1")
```

Comment:
If we sleep, why do we need this twice?
```
_ = eventsQueue.iterator().asScala.toList shouldBe List("Bar0", "Foo0", "Bar1", "Foo1")
```

Reply:
It makes sense to do it before the sleep because it should always hold at that point in time.
We do it after the sleep to verify that, even if we've waited for a bit, the state hasn't changed.

I we did it only after the sleep, the structure of the test could imply to the reader that we're waiting for some state to be reached. However, in reality, we're just making sure that state hasn't progressed

Reply:
you have convinced me